src/components/file-browser/file-browser-backends.ts

```diff
@@ -0,0 +1,131 @@
+import { FileBrowserDirectory, FileBrowserFile } from 'src/components/file-browser/file-browser-types'
+import { Ajax } from 'src/libs/ajax'
+
+
+interface RequestOptions {
+  signal?: AbortSignal
+}
+
+export interface IncrementalResponse<T> {
+  results: T[]
+  getNextPage(options?: RequestOptions): Promise<IncrementalResponse<T>>
+  hasNextPage: boolean
+}
+
+export interface FileBrowserBackend {
+  getDirectoriesInDirectory(path: string, options?: RequestOptions): Promise<IncrementalResponse<FileBrowserDirectory>>
+  getFilesInDirectory(path: string, options?: RequestOptions): Promise<IncrementalResponse<FileBrowserFile>>
+}
+
+interface GCSItem {
```

Comment:
Preferably, this would live in libs/ajax/GoogleStorage. However, I didn't want to convert that file to TS and I wasn't sure how to partially type the JS file via a .d.ts type declarations file.

Reply:
some thoughts on this:
1) let's at least move the backends files to libs/ajax to get them out of src/components which is a visual concerns area.
2) lets also create a libs/ajax/incremental-response.ts file to house the core types and such so that this pattern can be re-used for other instances where we want this incremental behavior.
3) maybe rename "backends" to "providers"?

Reply:
I acknowledge that these recommendations reflect an incremental "for now" state of things because the deeper consideration of code organization and more specific recommended patterns for separation-of-concerns and such is still a TODO piece of work.

Reply:
Added a partial type declaration for GoogleStorage.js in https://github.com/DataBiosphere/terra-ui/pull/3468/commits/40d8eabba74486057d2d394cac07c08819b69cc3.